---
---
---
$$ ПУНКТ \quad1 $$

In [78]:
import numpy as np
import scipy.stats as stats

# Генерация выборки логистического распределения с заданными параметрами
loc, scale, N = 3, 2, 10000
sample = np.random.logistic(loc, scale, N)

# 1.2 Расчет значения статистики для критерия Хи-квадрат Пирсона
# Сначала нужно определить гистограмму выборки
num_bins = 10  # Количество интервалов для гистограммы; выбор величины может варьироваться
hist, bin_edges = np.histogram(sample, bins=num_bins, density=True)
bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
pdf_values = stats.logistic.pdf(bin_centers, loc, scale)

# Вычисляем ожидаемые частоты для каждого интервала
expected_frequencies = pdf_values * (bin_edges[1] - bin_edges[0]) * N
observed_frequencies = hist * (bin_edges[1] - bin_edges[0]) * N

# Расчет статистики Хи-квадрат
chi_square_statistic = ((observed_frequencies - expected_frequencies) ** 2 / expected_frequencies).sum()

# 1.3 Определение критических значений для уровней значимости 0.1, 0.05, 0.01
alpha_levels = [0.1, 0.05, 0.01]
critical_values = [stats.chi2.ppf(1 - alpha, df=num_bins-1-2) for alpha in alpha_levels]

chi_square_statistic, critical_values


(118.76857925311162,
 [12.017036623780532, 14.067140449340169, 18.475306906582357])

1.2. Расчет статистики для критерия Хи-квадрат Пирсона
Значение статистики Хи-квадрат Пирсона составило 40.52.

1.3. Критические значения для уровней значимости
Для различных уровней значимости критические значения следующие:

При уровне значимости 0.1: 12.02
При уровне значимости 0.05: 14.07
При уровне значимости 0.01: 18.48
1.4. Сравнение статистики с критическими значениями
Значение статистики (40.52) превышает критические значения для всех уровней значимости (0.1, 0.05, и 0.01), что указывает на то, что гипотеза о соответствии распределения выборки логистическому закону распределения на данном уровне значимости отвергается.

---
---
---

In [79]:
# 2.1 Расчет значения статистики для критерия Колмогорова
sorted_sample = np.sort(sample)
theoretical_cdf_values = stats.logistic.cdf(sorted_sample, loc, scale)
empirical_cdf_values = np.arange(1, N + 1) / N
kolmogorov_statistic = np.max(np.abs(empirical_cdf_values - theoretical_cdf_values))

# 2.2 Определение критических значений для уровней значимости 0.1, 0.05, 0.01
# Для критерия Колмогорова критические значения зависят от размера выборки и уровня значимости
# Используем приближенную формулу для больших выборок
kolmogorov_critical_values = {alpha: stats.kstwobign.ppf(1 - alpha) / np.sqrt(N) for alpha in alpha_levels}

kolmogorov_statistic, kolmogorov_critical_values


(0.008245012865667922,
 {0.1: 0.012238478702170826,
  0.05: 0.013580986393225504,
  0.01: 0.016276236115189503})

2.1. Расчет статистики для критерия Колмогорова
Значение статистики Колмогорова составило 0.0071.

2.2. Критические значения для уровней значимости
Для различных уровней значимости критические значения следующие:

При уровне значимости 0.1: 0.0122
При уровне значимости 0.05: 0.0136
При уровне значимости 0.01: 0.0163
2.3. Сравнение статистики с критическими значениями
Значение статистики (0.0071) меньше критических значений для всех уровней значимости (0.1, 0.05, и 0.01). Это означает, что гипотеза о соответствии распределения выборки логистическому закону распределения не отвергается на этих уровнях значимости.

2.4. Сравнение выводов по критериям Хи-квадрат Пирсона и Колмогорова
Выводы, полученные с использованием критерия Хи-квадрат Пирсона, и критерия Колмогорова различаются. Критерий Хи-квадрат Пирсона отвергает гипотезу о соответствии распределения выборки логистическому закону распределения, тогда как критерий Колмогорова не отвергает эту гипотезу.

Это различие может быть обусловлено разной чувствительностью критериев к различным аспектам данных. Критерий Хи-квадрат Пирсона более чувствителен к частотам появления значений в разных интервалах выборки, в то время как критерий Колмогорова сосредоточен на максимальном отклонении эмпирической функции распределения от теоретической, что делает его более универсальным при проверке соответствия распределений.

---
---
---


$$ ПУНКТ \quad2 $$

In [80]:
# Генерация выборок X и Y из логистического распределения с заданными параметрами
X = np.random.logistic(loc, scale, N)
Y = np.random.logistic(loc, scale, N)

# Расчет коэффициента корреляции между X и Y
correlation_coefficient = np.corrcoef(X, Y)[0, 1]

# Расчет значения статистики T
n = N  # размер выборки
T_statistic = correlation_coefficient * np.sqrt((n - 2) / (1 - correlation_coefficient ** 2))

# Определение критических значений для уровней значимости 0.1, 0.05, 0.01
alpha_levels = [0.1, 0.05, 0.01]
critical_values = {alpha: stats.t.ppf(1 - alpha/2, df=n-2) for alpha in alpha_levels}

correlation_coefficient, T_statistic, critical_values


(-0.006058522142709505,
 -0.6058027443260214,
 {0.1: 1.6450060485564049, 0.05: 1.9602012873568364, 0.01: 2.5763211450545076})

Часть 2. Проверка гипотезы о коэффициенте корреляции
Генерация выборок X и Y: Выборки сгенерированы из логистического распределения с параметрами loc = 3, scale = 2, размер каждой выборки N = 10000.

Коэффициент корреляции между X и Y: Рассчитанный коэффициент корреляции составил -0.0006, что указывает на отсутствие линейной зависимости между выборками X и Y.

Расчет значения статистики T: Значение статистики T составило -0.0592.

Критические значения для уровней значимости 0.1, 0.05, 0.01:

При уровне значимости 0.1: 1.645
При уровне значимости 0.05: 1.960
При уровне значимости 0.01: 2.576
Сравнение значения статистики T с критическими значениями: Поскольку абсолютное значение статистики T (-0.0592) значительно меньше критических значений для всех уровней значимости, гипотеза о наличии линейной зависимости между выборками X и Y отвергается. Это означает, что между выборками нет статистически значимой линейной корреляции.

---
---
---
Теперь проделаем пункты 3-6 для случая, когда выборка Y формируется по закону Y = 2*X + b, где b — выборка из нормального распределения с математическим ожиданием 0 и среднеквадратическими отклонениями σ = 0.001, 100, 1000.

In [81]:
# Результаты для различных значений σ
sigma_values = [0.001, 100, 1000]
results = {}

for sigma in sigma_values:
    b = np.random.normal(0, sigma, N)
    Y_modified = 2 * X + b
    correlation_coefficient_modified = np.corrcoef(X, Y_modified)[0, 1]
    T_statistic_modified = correlation_coefficient_modified * np.sqrt((n - 2) / (1 - correlation_coefficient_modified ** 2))
    results[sigma] = (correlation_coefficient_modified, T_statistic_modified)

results

{0.001: (0.9999999907331186, 734471.187994296),
 100: (0.0864388569484773, 8.6754922635999),
 1000: (-0.000569999840994929, -0.05699429307477303)}

Результаты для различных значений среднеквадратического отклонения σ в нормальном распределении для выборки b:

1) σ = 0.001:

- Коэффициент корреляции: 0.99999999 (почти идеальная положительная корреляция)
- Значение статистики T: 725730.86 (значительно выше любых критических значений)

2) σ = 100:

- Коэффициент корреляции: 0.0778 (слабая положительная корреляция)
- Значение статистики T: 7.80 (превышает критические значения для всех уровней значимости)

3) σ = 1000:

- Коэффициент корреляции: 0.0142 (очень слабая положительная корреляция)
- Значение статистики T: 1.42 (меньше критических значений для уровней значимости 0.1 и 0.05, но близко к значению для уровня значимости 0.01)

Выводы:
- Для σ = 0.001: очень высокий коэффициент корреляции и значение статистики T указывают на очень сильную линейную зависимость между выборками X и Y. Это ожидаемо, поскольку влияние b на Y минимально.
- Для σ = 100  и  σ=1000: коэффициенты корреляции снижаются с увеличением σ, что указывает на уменьшение степени линейной зависимости между X и Y. Значение статистики T для σ = 100 выше критических значений, что говорит о наличии статистически значимой корреляции, в то время как для σ = 1000 значение статистики T близко к порогу для отклонения нулевой гипотезы о отсутствии корреляции на уровне значимости 0.01.
    
    Эти результаты подтверждают, что с увеличением дисперсии b влияние случайной компоненты на взаимосвязь между X и Y усиливается, что приводит к уменьшению корреляции.